src/shared/codelens/goCodeLensProvider.ts

```diff
@@ -0,0 +1,195 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { VSCODE_EXTENSION_ID } from '../extensions'
+import { LambdaHandlerCandidate } from '../lambdaHandlerSearch'
+import { getLogger } from '../logger/logger'
+import { findParentProjectFile } from '../utilities/workspaceUtils'
+
+export const GO_LANGUAGE = 'go'
+export const GO_ALLFILES: vscode.DocumentFilter[] = [
+    {
+        scheme: 'file',
+        language: GO_LANGUAGE,
+    },
+]
+
+// Go modules were introduced in Go 1.11
+// Before that, $GOPATH had to be set properly to find dependencies
+// We currently just ignore projects without a Go module file
+export const GO_BASE_PATTERN = '**/go.mod'
+
+export async function getLambdaHandlerCandidates(document: vscode.TextDocument): Promise<LambdaHandlerCandidate[]> {
+    const modFile: vscode.Uri | undefined = await findParentProjectFile(document.uri, /go\.mod$/)
+    const goIsActive: boolean = await checkForGoExtension()
+
+    if (!modFile || !goIsActive) {
+        return []
+    }
+
+    // We'll check that the Go lambda module is required, otherwise showing debug configs does not make much sense
+    // If we want to support GOPATH, then we should check for an import statemnt within the file instead
+    try {
+        const modDoc: vscode.TextDocument = await vscode.workspace.openTextDocument(modFile!)
+
+        if (
+            !modDoc.getText().includes('require github.com/aws/aws-lambda-go') ||
+            !document.getText().includes('github.com/aws/aws-lambda-go/lambda')
+        ) {
+            return []
+        }
+    } catch (err) {
+        // No need to throw an error
+        getLogger().verbose(
+            `Go CodeLens: not enabled for "${document.fileName}". Verify that a go.mod file exists and is within the module directory`
+        )
+
+        return []
+    }
+
+    const filename = document.uri.fsPath
+    const symbols: vscode.DocumentSymbol[] =
+        (await vscode.commands.executeCommand<vscode.DocumentSymbol[]>(
+            'vscode.executeDocumentSymbolProvider',
+            document.uri
+        )) ?? []
+
+    return symbols
+        .filter(symbol => isValidFuncSignature(document, symbol))
+        .map<LambdaHandlerCandidate>(symbol => {
+            return {
+                filename,
+                handlerName: `${document.uri}.${symbol.name}`,
+                manifestUri: modFile,
+                range: symbol.range,
+            }
+        })
+}
+
+/**
+ * Checks for a valid lamba function signature for Go
+ *
+ * @param document VS Code Document that contains the symbol
+ * @param symbol VS Code DocumentSymbol to analyze
+ */
+export function isValidFuncSignature(document: vscode.TextDocument, symbol: vscode.DocumentSymbol): boolean {
+    if (symbol.name === 'main') {
```

Comment:
Correct me if I'm wrong but all Lambda functionality is driven through the `main` function in Go? https://docs.aws.amazon.com/lambda/latest/dg/golang-handler.html

I'm wondering if it's right to even draw a codelens on non-main functions, since the main function determines what code is called? Also, how does this impact the handler name selection? Do we even need to show a symbol there? Does this also mean we have to check for `package main`?

(I know nothing about Go so feel free to put me in my place)

Reply:
You're correct, it is all through the `main` function. So realistically it only makes sense to draw a single CodeLens per package. The provider could be made to check that the `main` function body includes a `start` call to a valid handler. Then we will only draw a lens if there exists both a valid handler and a start call for it. Should the lens be placed over `main` or over the handler? Over the handler is more consistent with our other languages, but of course `main` makes more sense for Go.

The handler name can be anything and that currently does not matter for lens drawing. `package main` could be another required component since it is needed to be made into a binary.